libkbfs/disk_block_cache.go

```diff
@@ -619,60 +646,56 @@ func (cache *DiskBlockCacheStandard) Put(ctx context.Context, tlfID tlf.ID,
 				"Error writing to TLF cache database: %+v", err)
 		}
 	}
-	// Initially set HasPrefetched to false; rely on UpdateMetadata to fix it.
+	// Initially set HasPrefetched and DonePrefetch to false; rely on
+	// UpdateMetadata to fix it.
 	return cache.updateMetadataLocked(ctx, tlfID, blockKey, int(encodedLen),
-		false)
+		false, false)
+}
+
+// GetMetadata implements the DiskBlockCache interface for
+// DiskBlockCacheStandard.
+func (cache *DiskBlockCacheStandard) GetMetadata(ctx context.Context,
+	blockID kbfsblock.ID) (DiskBlockCacheMetadata, error) {
+	cache.shutdownLock.RLock()
+	defer cache.shutdownLock.RUnlock()
+	return cache.getMetadataLocked(blockID)
 }
 
 // UpdateMetadata implements the DiskBlockCache interface for
 // DiskBlockCacheStandard.
 func (cache *DiskBlockCacheStandard) UpdateMetadata(ctx context.Context,
-	blockID kbfsblock.ID, hasPrefetched bool) (err error) {
-	select {
-	case <-cache.startedCh:
-	default:
-		// If the cache hasn't started yet, return an error.
-		return DiskCacheStartingError{"UpdateMetadata"}
-	}
-	var md diskBlockCacheMetadata
-	// Only obtain a read lock because this happens on Get, not on Put.
-	cache.lock.RLock()
-	defer cache.lock.RUnlock()
-	// shutdownCh has to be checked under lock, otherwise we can race.
-	select {
-	case <-cache.shutdownCh:
-		return DiskCacheClosedError{"UpdateMetadata"}
-	default:
+	blockID kbfsblock.ID, hasPrefetched, donePrefetch bool) (err error) {
+
+	cache.shutdownLock.Lock()
+	defer cache.shutdownLock.Unlock()
+	err = cache.checkCacheLocked("UpdateMetadata")
+	if err != nil {
+		return errors.WithStack(err)
 	}
+
 	defer func() {
 		if err == nil {
 			cache.updateMeter.Mark(1)
 		}
 	}()
-	md, err = cache.getMetadata(blockID)
+	md, err := cache.getMetadataLocked(blockID)
 	if err != nil {
 		return NoSuchBlockError{blockID}
 	}
 	return cache.updateMetadataLocked(ctx, md.TlfID, blockID.Bytes(),
-		int(md.BlockSize), hasPrefetched)
+		int(md.BlockSize), hasPrefetched, donePrefetch)
 }
 
 // Size implements the DiskBlockCache interface for DiskBlockCacheStandard.
 func (cache *DiskBlockCacheStandard) Size() int64 {
-	select {
-	case <-cache.startedCh:
-	default:
-		// If the cache hasn't started yet, return 0.
-		return 0
-	}
-	cache.lock.RLock()
-	defer cache.lock.RUnlock()
-	// shutdownCh has to be checked under lock, otherwise we can race.
-	select {
-	case <-cache.shutdownCh:
+
```

Comment:
I'm using the blank line to separate out the locking and checking from the rest of the functionality for each of the interface methods.